// rise-foundation/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// S3
var get_file_url_1 = __importDefault(require("./s3/get_file_url"));
var get_file_1 = __importDefault(require("./s3/get_file"));
var upload_file_1 = __importDefault(require("./s3/upload_file"));
var remove_file_1 = __importDefault(require("./s3/remove_file"));
var bucket_1 = __importDefault(require("./s3/cf/bucket"));
var makeSimpleBucket_1 = __importDefault(require("./s3/cf/makeSimpleBucket"));
// DB
var db_1 = require("./db/db");
var makeDb_1 = require("./db/cf/makeDb");
// Cognito
var createUser_1 = require("./cognito/createUser");
var removeUser_1 = require("./cognito/removeUser");
var resetPassword_1 = require("./cognito/resetPassword");
var validateJwtToken_1 = require("./cognito/validateJwtToken");
var loginUser_1 = require("./cognito/loginUser");
var loginHandleNewPassword_1 = require("./cognito/loginHandleNewPassword");
var makeCognitoPoolAndClient_1 = require("./cognito/cf/makeCognitoPoolAndClient");
// Cloudformation
var deployStack_1 = require("./cloudformation/deployStack");
var getDeployStatus_1 = require("./cloudformation/getDeployStatus");
var getOutputs_1 = require("./cloudformation/getOutputs");
var getRemoveStatus_1 = require("./cloudformation/getRemoveStatus");
var removeStack_1 = require("./cloudformation/removeStack");
// Lambda
var updateLambdaCode_1 = require("./lambda/updateLambdaCode");
var makeLambda_1 = require("./lambda/cf/makeLambda");
var makeLambdaLayer_1 = require("./lambda/cf/makeLambdaLayer");
// CodeStar
var makeArtifactBucket_1 = require("./codestar/cf/makeArtifactBucket");
var makeGithubConnection_1 = require("./codestar/cf/makeGithubConnection");
var makeBuildProject_1 = require("./codestar/cf/makeBuildProject");
var makePipeline_1 = require("./codestar/cf/makePipeline");
var putJobSuccess_1 = require("./codestar/putJobSuccess");
var putJobFailure_1 = require("./codestar/putJobFailure");
exports.default = {
    s3: {
        cf: {
            makeBucket: bucket_1.default,
            makeSimpleBucket: makeSimpleBucket_1.default
        },
        uploadFile: upload_file_1.default,
        getFile: get_file_1.default,
        getFileUrl: get_file_url_1.default,
        removeFile: remove_file_1.default
    },
    db: {
        cf: {
            makeDb: makeDb_1.makeDb
        },
        get: db_1.getDbItem,
        list: db_1.listDbItems,
        set: db_1.setDbItem,
        remove: db_1.removeDbItem
    },
    cognito: {
        cf: {
            makeCognitoPoolAndClient: makeCognitoPoolAndClient_1.makeCognitoPoolAndClient
        },
        createUser: createUser_1.createUser,
        removeUser: removeUser_1.removeUser,
        resetPassword: resetPassword_1.resetPassword,
        validateToken: validateJwtToken_1.validateToken,
        loginUser: loginUser_1.loginUser,
        loginHandleNewPassword: loginHandleNewPassword_1.loginHandleNewPassword
    },
    cloudformation: {
        deployStack: deployStack_1.deployStack,
        getDeployStatus: getDeployStatus_1.getDeployStatus,
        getCloudFormationOutputs: getOutputs_1.getCloudFormationOutputs,
        removeStack: removeStack_1.removeStack,
        getRemoveStatus: getRemoveStatus_1.getRemoveStatus
    },
    lambda: {
        cf: {
            makeLambda: makeLambda_1.makeLambda,
            makeLambdaLayer: makeLambdaLayer_1.makeLambdaLayer
        },
        updateLambdaCode: updateLambdaCode_1.updateLambdaCode
    },
    codestar: {
        cf: {
            makeArtifactBucket: makeArtifactBucket_1.makeArtifactBucket,
            makeGithubConnection: makeGithubConnection_1.makeGithubConnection,
            makeBuildProject: makeBuildProject_1.makeBuildProject,
            makePipeline: makePipeline_1.makePipeline
        },
        putJobSuccess: putJobSuccess_1.putJobSuccess,
        putJobFailure: putJobFailure_1.putJobFailure
    }
};
