// rise-foundation/dist/cognito/cf/makeCognitoPoolAndClient.js
"use strict";
/*

    CognitoUserPoolMyUserPool:
        Type: AWS::Cognito::UserPool
        Properties:
            UserPoolName: RiseFoundationIntegrationTest-pool

            AdminCreateUserConfig:
                AllowAdminCreateUserOnly: true
                InviteMessageTemplate:
                    EmailSubject: You are being invited to join Grandvalley Strawberries
                UnusedAccountValidityDays: 365

    CognitoUserPoolClient:
        Type: AWS::Cognito::UserPoolClient
        Properties:
            ClientName: RiseFoundationIntegrationTest-pool-client
            UserPoolId:
                Ref: CognitoUserPoolMyUserPool
            ExplicitAuthFlows:
                - ADMIN_NO_SRP_AUTH
            GenerateSecret: false
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeCognitoPoolAndClient = void 0;
function makeCognitoPoolAndClient(name) {
    return {
        Resources: {
            CognitoUserPool: {
                Type: 'AWS::Cognito::UserPool',
                Properties: {
                    UserPoolName: name + "-pool",
                    AdminCreateUserConfig: {
                        AllowAdminCreateUserOnly: true,
                        UnusedAccountValidityDays: 365
                    }
                }
            },
            CognitoUserPoolClient: {
                Type: 'AWS::Cognito::UserPoolClient',
                Properties: {
                    ClientName: name + "-pool-client",
                    UserPoolId: {
                        Ref: 'CognitoUserPool'
                    },
                    ExplicitAuthFlows: ['ADMIN_NO_SRP_AUTH'],
                    GenerateSecret: false
                }
            }
        },
        Outputs: {
            UserPoolId: {
                Value: {
                    Ref: 'CognitoUserPool'
                }
            },
            UserPoolClientId: {
                Value: {
                    Ref: 'CognitoUserPoolClient'
                }
            }
        }
    };
}
exports.makeCognitoPoolAndClient = makeCognitoPoolAndClient;
