// rise-foundation/dist/s3/cf/bucket.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function makeBucket(name) {
    var _a, _b;
    var theName = name.charAt(0).toUpperCase() + name.slice(1);
    var BucketName = theName + "Bucket";
    var PolicyName = theName + "BucketPolicy";
    return {
        Resources: (_a = {},
            _a[BucketName] = {
                Type: 'AWS::S3::Bucket',
                Properties: {
                    BucketEncryption: {
                        ServerSideEncryptionConfiguration: [
                            {
                                ServerSideEncryptionByDefault: {
                                    SSEAlgorithm: 'AES256'
                                }
                            }
                        ]
                    }
                }
            },
            _a[PolicyName] = {
                Type: 'AWS::S3::BucketPolicy',
                Properties: {
                    Bucket: {
                        Ref: BucketName
                    },
                    PolicyDocument: {
                        Statement: [
                            {
                                Action: 's3:*',
                                Effect: 'Deny',
                                Principal: '*',
                                Resource: [
                                    {
                                        'Fn::Join': [
                                            '',
                                            [
                                                'arn:',
                                                {
                                                    Ref: 'AWS::Partition'
                                                },
                                                ':s3:::',
                                                {
                                                    Ref: BucketName
                                                },
                                                '/*'
                                            ]
                                        ]
                                    }
                                ],
                                Condition: {
                                    Bool: {
                                        'aws:SecureTransport': false
                                    }
                                }
                            }
                        ]
                    }
                }
            },
            _a),
        Outputs: (_b = {},
            _b[BucketName] = {
                Value: {
                    Ref: BucketName
                }
            },
            _b)
    };
}
exports.default = makeBucket;
